# Pet Adoption Prediction

The purpouse of this project is to predict the speed at which a pet is adopted, based on the data extracted from the pet’s listing on PetFinder.my (Malaysia’s leading online animal welfare platform) and available on [https://www.kaggle.com/c/petfinder-adoption-prediction/data].

The data includes text, tabular and image data for the pets. Exploratory data analysis can be used to derive relationships between the adoption speed and the various parameters available from the pet’s profile and suggest improvements to the profiles that would increase the animal’s adoptability. 
Some profiles represent a group of pets. In this case, the speed of adoption is determined by the speed at which all of the pets are adopted

## Data Fields
- PetID - Unique hash ID of pet profile
- AdoptionSpeed - Categorical speed of adoption. Lower is faster. This is the value to predict. See below section for more info.
- Type - Type of animal (1 = Dog, 2 = Cat)
- Name - Name of pet (Empty if not named)
- Age - Age of pet when listed, in months
- Breed1 - Primary breed of pet (Refer to BreedLabels dictionary)
- Breed2 - Secondary breed of pet, if pet is of mixed breed (Refer to BreedLabels dictionary)
- Gender - Gender of pet (1 = Male, 2 = Female, 3 = Mixed, if profile represents group of pets)
- Color1 - Color 1 of pet (Refer to ColorLabels dictionary)
- Color2 - Color 2 of pet (Refer to ColorLabels dictionary)
- Color3 - Color 3 of pet (Refer to ColorLabels dictionary)
- MaturitySize - Size at maturity (1 = Small, 2 = Medium, 3 = Large, 4 = Extra Large, 0 = Not Specified)
- FurLength - Fur length (1 = Short, 2 = Medium, 3 = Long, 0 = Not Specified)
- Vaccinated - Pet has been vaccinated (1 = Yes, 2 = No, 3 = Not Sure)
- Dewormed - Pet has been dewormed (1 = Yes, 2 = No, 3 = Not Sure)
- Sterilized - Pet has been spayed / neutered (1 = Yes, 2 = No, 3 = Not Sure)
- Health - Health Condition (1 = Healthy, 2 = Minor Injury, 3 = Serious Injury, 0 = Not Specified)
- Quantity - Number of pets represented in profile
- Fee - Adoption fee (0 = Free)
- State - State location in Malaysia (Refer to StateLabels dictionary)
- RescuerID - Unique hash ID of rescuer
- VideoAmt - Total uploaded videos for this pet
- PhotoAmt - Total uploaded photos for this pet
- Description - Profile write-up for this pet. The primary language used is English, with some in Malay or Chinese

In [1]:
#!/usr/bin/env python
# coding: utf-8

# ## Understanding the data

# Import pandas/os
import pandas as pd
import os

# Define files and directories
os.chdir('C:/Users/isado/Documents/SpringBoard/Capstone1/')
         
# Create DataFrame with training and testing data 
df_train = pd.read_csv('./train/train.csv')
df_test = pd.read_csv('./test/test.csv')

# First look at imported data: select sample of 3 random rows 
df_train.sample(n=3)


Type             Name  Age  Breed1  Breed2  Gender  Color1  Color2  \
2633      2  The Catdashians   11     266     266       3       1       3   
1774      1            Harry    1     206     189       1       1       5   
13942     2              NaN    2     266       0       2       1       2   

       Color3  MaturitySize      ...        Health  Quantity  Fee  State  \
2633        7             2      ...             1         6    0  41326   
1774        0             1      ...             1         1    0  41326   
13942       6             1      ...             1         1    0  41326   

                              RescuerID  VideoAmt  \
2633   5fc6b4f39959a3d129490777ecaf572b         0   
1774   966edbab6abf83106b0b9fdde1922975         0   
13942  f785961a3984e8adc7bf95c73e1c2544         0   

                                             Description      PetID PhotoAmt  \
2633   Hi, if you are looking for a loving pet and ha...  058d37dda      5.0   
1774   Harry is among 5 siblings who is looking for a...  5bc5e3faa      1.0   
13942  Adorable, Playful & Loving She's so cute, manj...  8cb79284e      2.0   

       AdoptionSpeed  
2633               4  
1774               1  
13942              3  

[3 rows x 24 columns]

In [2]:
# Print DataFrame summary
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14993 entries, 0 to 14992
Data columns (total 24 columns):
Type             14993 non-null int64
Name             13736 non-null object
Age              14993 non-null int64
Breed1           14993 non-null int64
Breed2           14993 non-null int64
Gender           14993 non-null int64
Color1           14993 non-null int64
Color2           14993 non-null int64
Color3           14993 non-null int64
MaturitySize     14993 non-null int64
FurLength        14993 non-null int64
Vaccinated       14993 non-null int64
Dewormed         14993 non-null int64
Sterilized       14993 non-null int64
Health           14993 non-null int64
Quantity         14993 non-null int64
Fee              14993 non-null int64
State            14993 non-null int64
RescuerID        14993 non-null object
VideoAmt         14993 non-null int64
Description      14981 non-null object
PetID            14993 non-null object
PhotoAmt         14993 non-null float64
AdoptionSpe

In [3]:
# Print DataFrame summary
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3972 entries, 0 to 3971
Data columns (total 23 columns):
Type            3972 non-null int64
Name            3561 non-null object
Age             3972 non-null int64
Breed1          3972 non-null int64
Breed2          3972 non-null int64
Gender          3972 non-null int64
Color1          3972 non-null int64
Color2          3972 non-null int64
Color3          3972 non-null int64
MaturitySize    3972 non-null int64
FurLength       3972 non-null int64
Vaccinated      3972 non-null int64
Dewormed        3972 non-null int64
Sterilized      3972 non-null int64
Health          3972 non-null int64
Quantity        3972 non-null int64
Fee             3972 non-null int64
State           3972 non-null int64
RescuerID       3972 non-null object
VideoAmt        3972 non-null int64
Description     3971 non-null object
PetID           3972 non-null object
PhotoAmt        3972 non-null float64
dtypes: float64(1), int64(18), object(4)
memory usage: 713.

# Check for missing values

In [4]:
#Function that calculates percentage of missing data
def missingData(dataframe):
    
    for item in (dataframe.isnull().sum()).iteritems():
        if item[1] > 0:
            print('Missing Data percentage for '+item[0]+' is {:2.2%}'.format((item[1]/dataframe.shape[0])) )
        
missingData(df_train)

Missing Data percentage for Name is 8.38%
Missing Data percentage for Description is 0.08%


In [5]:
missingData(df_test)

Missing Data percentage for Name is 10.35%
Missing Data percentage for Description is 0.03%


# Cross Tabulation

In [6]:
df_train.columns

Index(['Type', 'Name', 'Age', 'Breed1', 'Breed2', 'Gender', 'Color1', 'Color2',
       'Color3', 'MaturitySize', 'FurLength', 'Vaccinated', 'Dewormed',
       'Sterilized', 'Health', 'Quantity', 'Fee', 'State', 'RescuerID',
       'VideoAmt', 'Description', 'PetID', 'PhotoAmt', 'AdoptionSpeed'],
      dtype='object')

In [7]:
pd.crosstab(df_train.AdoptionSpeed, df_train.Type)#1 is Dog, 2 is Cat

Type              1     2
AdoptionSpeed            
0               170   240
1              1435  1655
2              2164  1873
3              1949  1310
4              2414  1783

In [8]:
pd.crosstab(df_train.AdoptionSpeed, df_train.Age)

Age            0    1     2    3    4    5    6    7    8    9   ...   132  \
AdoptionSpeed                                                    ...         
0               10   54   115   44   24    7   20    7    4    5 ...     0   
1               50  643   865  408  211  100   80   43   36   24 ...     1   
2               44  754  1120  586  265  157  117   59   68   51 ...     1   
3               39  511   783  458  260  135  115   62   70   31 ...     3   
4               36  342   620  470  349  196  226  110  131   73 ...     3   

Age            135  144  147  156  168  180  212  238  255  
AdoptionSpeed                                               
0                0    0    0    0    0    0    0    0    0  
1                0    0    1    0    0    0    0    0    0  
2                0    1    0    1    0    0    0    0    0  
3                1    1    0    0    1    1    2    0    0  
4                0    2    0    0    0    1    1    1    2  

[5 rows x 106 columns]

With 106 columns it might be beneficial to convert the pet's age into years, rounding to the nearest .5. 
255 months seems pretty old at 21 years and change, but with the oldest recorded cat at 38 yo and the oldest recorded dog at 29yo, so we cannot rule it out.


In [9]:
pd.crosstab(df_train.AdoptionSpeed, df_train.Health)

Health            1    2   3
AdoptionSpeed               
0               392   17   1
1              2999   89   2
2              3925  106   6
3              3150   98  11
4              4012  171  14